### PR TITLE
implement extract_graph_relationships + tests + refactor

### PR DESCRIPTION
**Pull request by RitxmSaha**
Apologies for the big diff. Most of it is from refactoring/deleting, I'll detail whats important.

**lib/sycamore/sycamore/docset.py**

- Added all the transforms + comments here

**lib/sycamore/sycamore/extract_graph_relationships**
- Extracted relationships from nodes within sections
- Added Enum's to pydantic schemas to ensure LLM chooses only from the list of entities of that specific type

**lib/sycamore/sycamore/resolve_graph_entities**

- Added resolving duplicate relationships when merging nodes

### CODE REVIEW

File: lib/sycamore/sycamore/transforms/extract_document_structure.py [hunk 1]
```diff
@@ -114,3 +114,29 @@ def __init__(
         **resource_args,
     ):
         super().__init__(child, f=structure.extract, **resource_args)
+
```
Comment (RitxmSaha) [inline lib/sycamore/sycamore/transforms/extract_document_structure.py:117]: this code was just moved from extract_graph.py

File: lib/sycamore/sycamore/transforms/extract_graph_entities.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/transforms/extract_graph_entities.py:1]: removed metadata extractor, this class only extracts entities! future framework to add unsupervised entity extractor

Comment (RitxmSaha) [reply lib/sycamore/sycamore/transforms/extract_graph_entities.py:1]: most code was ported over from extract_graph.py

File: lib/sycamore/sycamore/transforms/extract_graph_relationships.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/transforms/extract_graph_relationships.py:1]: This is new, same general framework as extract_graph_entities.py though.

File: lib/sycamore/sycamore/tests/unit/transforms/test_graph_entity_extractor.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/tests/unit/transforms/test_graph_entity_extractor.py:1]: just a test, mostly ported over from test_graph_extractor.py

File: lib/sycamore/sycamore/tests/integration/connectors/neo4j/test_docset_to_neo4j.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/tests/integration/connectors/neo4j/test_docset_to_neo4j.py:1]: changed this integration test to work on a doc with actual entities/relationships.

File: lib/sycamore/sycamore/tests/unit/transforms/test_graph_relationship_extractor.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/tests/unit/transforms/test_graph_relationship_extractor.py:1]: similar test as test_graph_entity_extractor, not much attention needed

File: lib/sycamore/sycamore/llms/openai.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/llms/openai.py:1]: type checking change

File: lib/sycamore/sycamore/llms/llms.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/llms/llms.py:1]: type checking change

File: lib/sycamore/sycamore/docset.py
Comment (RitxmSaha) [inline lib/sycamore/sycamore/docset.py:1]: added new transforms + new documentation

File: lib/sycamore/sycamore/transforms/extract_graph_relationships.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        # Use mypy ignore type since pydantic has bad interaction with mypy with creating class from a variable class
+        fields = {relation.__name__: (List[relation], ...) for relation in parsed_relations}  # type: ignore
+        relationships_model = create_model("relationships", __base__=BaseModel, **fields)  # type: ignore
+
+        entities = ""
+        for key, nodes in parsed_nodes.items():
+            entities += f"{key}:\n"
+            for node in nodes:
+                entities += f"{node}\n"
```
Comment (HenryL27) [inline lib/sycamore/sycamore/transforms/extract_graph_relationships.py:138]: O(n) more efficient to build up a list of strings and then join them once at the end (bc strings are immutable, so string += "hi" is a linear time op)

**Review by HenryL27 — APPROVED** [review]
couple nits. The string thing matters. Don't worry about the extract_document_structre naming thing.

**Pull request merged**